lib/spack/spack/solver/concretize.lp

```diff
@@ -472,6 +472,21 @@ error(2, "Attempted to use external for '{0}' which does not satisfy any configu
 #defined external_spec_condition/4.
 #defined external_spec_condition/5.
 
+%-----------------------------------------------------------------------------
+% Config required semantics
+%-----------------------------------------------------------------------------
+
+% A package may have an associated "condition_group": a set of specs where
+% at least one must be applied. Each spec in the group is associated
+% with a condition via asp.impose(), and we require that exactly one condition
+% from each condition_group holds.
+1 { condition_holds(Y) : condition_group_member(Y, X) } 1 :- node(Package), condition_group(Package, X).
```

Comment:
I thought replacing this with

```
condition_group_satisfied(X) :- node(Package), condition_group(Package, X), condition_group_member(Y, X), condition_holds(Y).

error(2, "Cannot satisfy condition group for package '{0}'", Package)
  :- node(Package), condition_group(Package, X), not condition_group_satisfied(X).
```

Would fit better (and I could also pass along more values for better error messages) but if I do

```
$ spack -C test-cfgs/example1/ spec openmpi
==> Error: Package condition group error for openmpi
```

(i.e. now it fails: with the current `concretize.lp` logic it succeeds).

Reply:
This looks analogous to `node_os`: instead of removing

```
1 { condition_holds(Y) : condition_group_member(Y, X) } 1 :- node(Package), condition_group(Package, X).
```

I have relaxed it to

```
{ condition_holds(Y) : condition_group_member(Y, X) } :- node(Package), condition_group(Package, X).
```

I'm not sure why the `error :- ...` is needed though for cases where no condition group members apply (and likewise for OS choice).

Reply:
I think understand this better after reading [here](https://www.cs.cmu.edu/~cmartens/asp-notes.pdf): the relaxed choice rule allows no assignment; the error rule forces a choice where possible (but, in keeping with the other constraints, allows generating an error fact when that is not possible).